Draft: Added User defined states

BernhardBln opened the pull request:
Hi,
I mentioned earlier I needed user-defined states from the Bosch world as Switches in the Homekit world.

This is a draft. I am really not familiar with typescript (just with java), but I managed to get it working.
To be more precise, only after I updated the dependencies, I got errors about Buffer not being generic, which went away after the update. Probably connected to my local setup...

I would have loved to not have those extra files for user-defined states like stateSwitch.ts, but rather include them in accessory.ts, but I felt like I couldn't get the states into the data structures you are using for the climate controls.

What's your thoughts? If you are fine with integrating this into your plugin, I'd definitely suggest to add the following config options:

- disable climate controls (default: false), to avoid double entries if you already use the direct bosch integration
- disable user-defined states (default: true), so that people that use this already are not flooded with switches in Homekit for their states if they don't need it.

If not, I'd consider publishing my own plugin, based on this, but only with the states and not with the climate controls, but I think it would make sense to bundle this.

I also think I found and fixed some bugs, check the comments.

BernhardBln committed:
first attempt for user defined states

BernhardBln committed:
fixes to version

BernhardBln committed:
more fixes

BernhardBln commented on src/bshcApi.ts:
```diff
@@ -69,13 +71,42 @@ export class BshcApi {
+        .getBshcClient()
+        .getUserDefinedStates(deviceId)
+        .pipe(
+          map((response: BshbResponse<any>) => response.parsedResponse as BoschUserDefinedState))
```
this was weird. getUserDefinedStates(deviceId) claims to return an array, but in fact it was just a single element, which makes sense because this is what GET /states/id does, while GET /states returns an array.

fabiandev replied:
Is it a problem with the API, or the API client? Maybe leave a comment with this info in the code

BernhardBln committed:
fixed linting

BernhardBln commented on src/platform.ts:
```diff
@@ -257,17 +292,63 @@ export class BoschRoomClimateControlPlatform implements DynamicPlatformPlugin {
 
     const controller = new BoschRoomClimateControlAccessory(this, accessory);
     this.controllers.push(controller);
+    this.accessories.push(accessory);
```
I believe this was a bug, it just didn't cause any problems because you do not react on delete events for a climate controller (if they even exist?)

I did react on the deleted flag for userDefinedEvents, and with the same code as yours, I ran into the problem that the corresponding list for states did not contain anything.

If you want to react on deletion events in the future, you probably want this. If not, it will work for now.

fabiandev replied:
Interesting, I cannot recall to be honest. There is a `configureAccessory` function which does nothing else than push the reference to the array.

Maybe it's called at some point in the flow, but I may also be wrong.

fabiandev replied:
Would be a nice improvement though to react on deleted flag. Could this potentially remove the need for `periodicAccessorySync`?

BernhardBln replied:
Unfortunately, climate controls don't seem to have a deleted flag. It only comes with state updates. Let me check if there are some other notifications for that.

Regarding `configureAccessory`, the docs say its called when restoring from the cache, in case of a restart for example, and that was also my experience, I had added a state, restarted the bridge, added another, and when deleting the first it succeeded, while deleting the second failed as it wasn't in the collection.

BernhardBln replied:
Hm, the Thermosthat II seems to have this flag, the previous version (which I have) seems not to have it, I also think I didn't see it in the events that came when I pulled the events with postman.

So if you remove periodicAccessorySync, it would even be more important to be able to disable the climate controls like I suggested, or just do it when there is a first-gen control?

https://github.com/BoschSmartHome/bosch-shc-api-docs/blob/20ef514114d5be46da115f7d3bc5746e3cf0b8ff/openapi/Thermostat-II-local-openapi-v3.yml

BernhardBln replied:
But I can try to remove it for the userDefinedStates.

BernhardBln replied:
Tried and tested, works like a charm.

BernhardBln replied:
Ok, let me summarise the suggestion:

- periodic sync for state updates: 
  - turned off for user-defined states, long polling works great
  - still turned on for climate control
- periodic sync for devices (in platform.ts):
  - turned off for user-defined states, I use the deleted flag there
  - still turned on for climate controls, could be disabled if no first-gen controls are there, or if climate controls would be disabled all together (my personal use case)

So if I for example only wanted to use the user-defined states, there should be no periodic sync at all.

BernhardBln commented on src/platform.ts:
```diff
@@ -257,17 +292,63 @@ export class BoschRoomClimateControlPlatform implements DynamicPlatformPlugin {
 
     accessory.context.device = device;
     accessory.context.room = room;
+    accessory.displayName = this.getAccessoryDisplayName(room.name, device.name);
```
Guess that was missing.

fabiandev replied:
Not sure if this is needed here? Looks like the display name is set when creating the accessory a few lines above

BernhardBln committed:
removed comment

BernhardBln commented on src/platform.ts:
```diff
@@ -415,28 +563,62 @@ export class BoschRoomClimateControlPlatform implements DynamicPlatformPlugin {
     }
 
-    this.handleLongPollingResult(result);
+    await this.handleLongPollingResult(result);
```
I believe you want await here, as otherwise, modifications you are doing right now might interfere with modifications from the next events?

fabiandev replied:
`handleLongPollingResult` wasn't `async` before, meaning it should have been executed synchronously before as well.

BernhardBln replied:
True 😅

fabiandev reviewed (commented):
@BernhardBln thank you for contributing, appreciate your efforts and glad to hear you could get it up and running for your use-case! 💯

I'm not sure if we should put entities unrelated to climate control in this "climate control" plugin and provide an option to disable climate control accessories, it kind of defeats the purpose of this plugin for me 😅

On the other hand it would be nice to extend functionality – in this case maybe we can think about a good architecture to make this plugin extendable 🧠

I just addressed your comments for now, it's been a while since I created this plugin and unfortunately I don't have a Bosch Smart Home setup anymore for testing and development.

fabiandev commented:
@BernhardBln another thought: does the official Bosch HomeKit integration expose the user defined states? If yes, you could run this plugin for climate controls and the official integration in parallel, it shouldn't cause any issues.

BernhardBln committed:
from review

BernhardBln commented:
> @BernhardBln another thought: does the official Bosch HomeKit integration expose the user defined states? If yes, you could run this plugin for climate controls and the official integration in parallel, it shouldn't cause any issues.

No, that's why I came here in the first place :D 

I actually wanted to control the auto/manual state of my controls with user-defined states, mainly setting "nobody is at home" when everyone left the house (which Homekit can track, but the Bosch App cannot).

Since I found no way to connect the user-defined states with a "everyone left home" automation in Homekit, I thought to use your plugin to at  least toggle between automatic and manual mode when everyone left the house (like, when you are gone for two days, you woudn't want the heating start again in the first morning when nobody is at home).

I could of course set them all to manually and only control the temperatures throughout the day with Homekit, but I find the Bosch app nicer for those things, with the weekly schedules etc.

Then I found that the user-defined states are exposed in the api, so I thought I'd add them to your plugin.

At the same time, I would prefer to use the official integration for the climate controls, even if they are less flexible, just to be sure they will keep working in the future.  That's why in my case (especially still using the first-gen controls), I'd like to turn the climate-control part off, but that's my personal use case :)

I just feel like a plugin to only expose the states is too much, they would fit in here well, but at the same time I personally would prefer to stick with the original integration for the climate controls, and hence only use the states.

BernhardBln committed:
Config options and bugfix

- Added config options to enable user-defined states,
  and for disabling climate controls, defaults are:
  - user-defined switches disabled by default
  - climate-controls enabled by default
- bugfix: the code that should remove controls or states
  that do not exist any more was buggy

BernhardBln commented:
@fabiandev I made a proposal for the configuration, like that the plugin is 100% backwards compatible.

I also found and fixed another bug, you were iterating a list of controls to check if you need to remove or update them, and modified the collection while iterating over it, resulting in all controls not getting checked that were directly after a control that got removed.

Would you feel comfortable merging this? It will 
- not change anything for existing users
- but give them the bonus that they can expose user-defined states as switches in the home app if they want to (doing things like arming the alarm system in the Bosch app automatically when the last person left the house)
- allows me as a user who has first-gen climate controls to use the default implementation from Bosch (and not have the controls in Homekit twice)

BernhardBln committed:
updated readme

BernhardBln committed:
removed dist.tgz

BernhardBln committed:
marked as hb 2.0 ready

BernhardBln commented:
Also marked it as hb 2.0 ready. It worked with the states, with the 2nd gen controls I cannot test it as I don't have them.
